### PR TITLE
Allows explicit include/exclude of namespaces on restores

### LINKED ISSUE #22: add --exclude-namespaces to restore
It's nice to always backup a superset of namespaces but then restore everything instead a particular one or two. For example:

`ark restore create everything --exclude-namespaces=kube-system --restore-volumes`

It would be nice to have `--exclude-namespaces` on the restore option.

### PR DESCRIPTION
**Pull request by jrnt30**
Fixes #22

- Introduces similar Include/Exclude declaration on the Restore
resource and cli flags
- Removed the Namespace attribute
- Fixed a few small issues with the backup validation

Signed-off-by: Justin Nauman <justin.r.nauman@gmail.com>

### CODE REVIEW

**Comment by ncdc** [comment]
ok to test

File: pkg/apis/ark/v1/restore.go [hunk 1]
```diff
@@ -24,9 +24,19 @@ type RestoreSpec struct {
...
+	// should be used instead
 	// Namespaces is a slice of namespaces in the Ark backup to restore.
 	Namespaces []string `json:"namespaces"`
 
+	// IncludedNamespaces is a slice of namespace names to include objects
+	// from. If empty, all namespaces are included.
+	IncludedNamespaces []string `json:"includedNamespaces"`
+
+	// ExcludedNamespaces contains a list of namespaces that are not
+	// included in the backup.
```
Comment (skriss) [inline pkg/apis/ark/v1/restore.go]: replace "backup" with "restore"

File: pkg/cmd/cli/restore/create.go [hunk 2]
```diff
@@ -57,23 +57,26 @@ type CreateOptions struct {
...
+		IncludeNamespaces: flag.NewStringArray("*"),
 		NamespaceMappings: flag.NewMap().WithEntryDelimiter(",").WithKeyValueDelimiter(":"),
 		RestoreVolumes:    flag.NewOptionalBool(nil),
 	}
 }
 
 func (o *CreateOptions) BindFlags(flags *pflag.FlagSet) {
-	flags.Var(&o.Labels, "labels", "labels to apply to the restore")
-	flags.Var(&o.Namespaces, "namespaces", "comma-separated list of namespaces to restore")
+	flags.Var(&o.IncludeNamespaces, "include-namespaces", "namespaces to include in the backup (use '*' for all namespaces)")
```
Comment (skriss) [inline pkg/cmd/cli/restore/create.go]: replace "backup" with "restore"

File: pkg/cmd/cli/restore/create.go [hunk 3]
```diff
@@ -57,23 +57,26 @@ type CreateOptions struct {
...
+	flags.Var(&o.ExcludeNamespaces, "exclude-namespaces", "namespaces to exclude from the backup")
```
Comment (skriss) [inline pkg/cmd/cli/restore/create.go]: replace "backup" with "restore"

File: pkg/controller/restore_controller.go [hunk 4]
```diff
@@ -278,6 +285,10 @@ func (controller *restoreController) getValidationErrors(itm *api.Restore) []str
 		validationErrors = append(validationErrors, "BackupName must be non-empty and correspond to the name of a backup in object storage.")
 	}
```
Comment (skriss) [inline pkg/controller/restore_controller.go]: You should validate the the included/excluded namespaces are a valid set of includes/excludes. See [backup controller](https://github.com/heptio/ark/blob/master/pkg/controller/backup_controller.go#L299)

File: pkg/controller/restore_controller.go [hunk 5]
```diff
@@ -278,6 +285,10 @@ func (controller *restoreController) getValidationErrors(itm *api.Restore) []str
...
 
+	if len(itm.Spec.Namespaces) > 0 && len(itm.Spec.IncludedNamespaces) > 0 {
+		validationErrors = append(validationErrors, "Namespace and ItemNamespaces can not both be defined on the backup spec.")
```
Comment (skriss) [inline pkg/controller/restore_controller.go]: "Namespaces and IncludedNamespaces cannot both be defined on the restore spec."

**Comment by jrnt30** [comment]
Thanks for the feedback, I was curious about that as well.  The one thing that I hadn't initially thought of is that anyone that has Ark installed then would need to upgrade.  

I have seen project, specifically Helm, that have a tight coupling with Server vs. Client version.  While it's somewhat frustrating at times, I do see how it would significantly reduce the complexity over time until the API reaches a point of stability.

**Comment by jrnt30** [comment]
So having a bit more time to look at this properly my tests are failing due to some of the complexities around the handling of the migration of Namespace and the new namespace validation piece.  I will wait to hear if we need to support the retro attribute and if not just rework this a bit. 

Also as a note, the Jenkins auth doesn't allow viewing of the job currently even with the Github OAuth authorized.

**Comment by ncdc** [comment]
I'm fine removing the old flag. We'll make sure to include the required UX change in the release notes/changelog. Everyone ok with that?

**Comment by skriss** [comment]
Works for me.

**Comment by ncdc** [comment]
@jrnt30 please go ahead and remove the old flag - thanks!

File: pkg/controller/backup_controller.go [hunk 6]
```diff
@@ -292,11 +292,11 @@ func cloneBackup(in interface{}) (*api.Backup, error) {
 func (controller *backupController) getValidationErrors(itm *api.Backup) []string {
 	var validationErrors []string
 
-	for err := range collections.ValidateIncludesExcludes(itm.Spec.IncludedResources, itm.Spec.ExcludedResources) {
+	for _, err := range collections.ValidateIncludesExcludes(itm.Spec.IncludedResources, itm.Spec.ExcludedResources) {
```
Comment (skriss) [inline pkg/controller/backup_controller.go]: @jrnt30 thanks for catching!

**Comment by skriss** [comment]
updates LGTM, thanks @jrnt30 !

**Comment by skriss** [comment]
@ncdc pls review/merge if you approve.

**Pull request merged**